### PR TITLE
Feature/dapp stream

### PR DESCRIPTION
**Pull request by hunght**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env node
 const program = require('commander');
 const { prompt } = require('inquirer');
-const { newSigningKey, signingKeyChangePW } = require('./src/cliActions');
-const { newSigningKeySchema, changePasswordSigningKeySchema } = require('./src/promtSchema');
+const { newSigningKey, signingKeyChangePW, dappBuildBundleFile } = require('./src/cliActions');
+const { newSigningKeySchema, changePasswordSigningKeySchema, buildBundleFileSchema } = require('./src/promtSchema');
```
Comment (florianlenz) [inline index.js]: `buildBundleFileSchema` should be named `dappStreamSchema` think about it, you are not bundling the app. You are streaming it to a specific location

File: index.js [hunk 2]
```diff
@@ -29,6 +29,19 @@ program
             .catch((err) => console.error(err.message))
     });
 
+program
+  .command('dapp:build <signing-key-file>')
```
Comment (florianlenz) [inline index.js]: the issue says this is about streaming, not just building ....

File: index.js [hunk 3]
```diff
@@ -29,6 +29,19 @@ program
...
+  .option('-d, --dev', 'Dev mode build')
```
Comment (florianlenz) [inline index.js]: the name should be `Development build mode`

File: index.js [hunk 4]
```diff
@@ -29,6 +29,19 @@ program
...
+  .description('Build a DApp to an installable bundle')
```
Comment (florianlenz) [inline index.js]: thats not correct. This branch is about streaming it. Not about building a stand alone version.

File: index.js [hunk 5]
```diff
@@ -29,6 +29,19 @@ program
...
+  .action((signingKeyFile, cmd) => {
+    prompt(buildBundleFileSchema)
+        .then((answers) => {
+            return dappBuildBundleFile(answers, signingKeyFile, cmd.dev === true)
```
Comment (florianlenz) [inline index.js]: function name is incorrect. It's still about streaming.

File: src/cliActions.js [hunk 6]
```diff
@@ -112,7 +113,37 @@ const signingKeyChangePW = ({oldPassword, newPw, newPwConfirmation}, signingKeyF
 
 });
 
+/**
+ * @desc Build bundle json file
```
Comment (florianlenz) [inline src/cliActions.js]: the description is wrong, this is just about to stream the DApp

File: src/cliActions.js [hunk 7]
```diff
@@ -112,7 +113,37 @@ const signingKeyChangePW = ({oldPassword, newPw, newPwConfirmation}, signingKeyF
...
+    // make sure singing key exist
+    if (!fs.existsSync(signingKeyFile)) {
+      return rej(new Error(`Signing key ("${signingKeyFile}") does not exist`));
+    }
+
+    // read signing key
+    const rawSigningKey = fs.readFileSync(signingKeyFile, 'utf8');
+    const signingKey = JSON.parse(rawSigningKey);
+
+    // first decrypt the old signing key
```
Comment (florianlenz) [inline src/cliActions.js]: `// first decrypt the old signing key` looks like a copy / past mistake from the change password command

File: src/cliActions.test.js [hunk 8]
```diff
@@ -113,6 +115,27 @@ describe('cliActions', () => {
 
     });
 
+    describe('dappStreaming', () => {
```
Comment (florianlenz) [inline src/cliActions.test.js:119]: you should also have a case for if decrypting fails

File: src/promtSchema.js [hunk 9]
```diff
@@ -34,7 +34,16 @@ const changePasswordSigningKeySchema = [
     },
 ];
 
+const dappStreamingSchema = [
+    {
+        type: 'password',
+        name: 'pw',
+        message: 'Verify password of signing key ...',
```
Comment (florianlenz) [inline src/promtSchema.js]: `'Verify password of signing key ...'` makese no sense to me. you don't verify it, you just enter it.

File: src/webpack.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
+const path = require('path');
+const webpack = require('webpack');
+const fs = require('fs');
+
+/* Helper functions start*/
+const base64Encode = file => {
```
Comment (florianlenz) [inline src/webpack.js]: missing js doc

File: src/webpack.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  // read binary data
+  const bitmap = fs.readFileSync(file);
+  // convert binary data to base64 encoded string
+  return new Buffer(bitmap).toString('base64');
+};
+
+const checkFileExist = path => {
+  if (fs.existsSync(path)) {
+    return true;
+  } else {
```
Comment (florianlenz) [inline src/webpack.js]: unnecessary else and missing docs

Comment (florianlenz) [reply src/webpack.js]: actually the whole function makese no sense. `existsSync` returns an bool ...

File: src/webpack.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    console.log(`File not found at ${path}`);
+    return false;
+  }
+};
+/* Helper functions end*/
+
+const getDappMetaData = () => {
```
Comment (florianlenz) [inline src/webpack.js]: add js doc

File: src/webpack.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    : {};
+};
+
+
+
+/**
+ * Get data from webpack build bundle file and combine to metadata object
+ *  @param {function} callback return data with format {result:'',error:''}
+ */
+const getBuildObjectFromBundle = callback => {
```
Comment (florianlenz) [inline src/webpack.js]: please create correct JS doc

File: src/cliActions.js (hunk 7 above)
Comment (hunght) [reply src/cliActions.js]: done

File: src/promtSchema.js (hunk 9 above)
Comment (hunght) [reply src/promtSchema.js]: done

File: src/cliActions.test.js (hunk 8 above)
Comment (hunght) [reply src/cliActions.test.js:119]: done

File: src/webpack.js (hunk 11 above)
Comment (hunght) [reply src/webpack.js]: this function does 2 things, check existing file, and prompt error when file not found, I have renamed it to  checkFileExistAndPromptError

File: src/webpack.js (hunk 12 above)
Comment (hunght) [reply src/webpack.js]: done

File: src/webpack.js [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        },
+      });
+    } else {
+      callback({ error: err });
+    }
+  });
+};
+
+/**
+ * Watching webpack build process
```
Comment (florianlenz) [inline src/webpack.js]: missing `@desc`

**Review by florianlenz — CHANGES_REQUESTED** [review]

**Pull request merged**